spec/users_controller_spec.rb

```diff
@@ -0,0 +1,96 @@
+require_relative 'spec_helper'
+
+def app
+  UsersController.new
+end
+
+describe UsersController do
+  before do
+    @users = DB[:users]
+  end
+ 
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as an admin" do
+    before do
+      @admin = @users.where(admin: true).first
+      env "rack.session", {:current_user_id => @admin[:id]}
+    end
+
+    it "able to access '/users' page" do
+      get '/users'
+      last_response.body.must_include "Complete CRUD Users"
+    end
+
+    it "able to access other user's page" do
+      different_user_id = @admin[:id] + 1
+      get "/users/#{different_user_id}"
+      last_response.body.must_include "User Info"
+    end
+
+    it "displays all users '/users' page" do
+      number_of_users = @users.count
+      get '/users'
+      users_page = Nokogiri::HTML(last_response.body)
+      li_size = users_page.search("li").size
+      li_size.must_equal number_of_users
+    end
+  end
+
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as a regular user" do
+    before do
+      @regular_user = @users.where(admin: false).first
+      env "rack.session", {:current_user_id => @regular_user[:id]}
+    end
+
+    it "can view own page" do
+      get "/users/#{@regular_user[:id]}"
+      last_response.body.must_include "User Info"
+    end
+
+    it "redirects regular_user when accessing '/users' page" do
+      get '/users'
+      follow_redirect!
+      last_response.body.must_include "Unauthorized access"
```

Comment:
Instead of redirecting user, let's return the HTTP status code 403 here. Let's do this for all unauthorized access. https://en.wikipedia.org/wiki/HTTP_403